lib/tree.rb

```diff
@@ -2,63 +2,130 @@ class TreeNode
+
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+
+    return current_node
+  end
+
+  # Time Complexity: O(h) where h is the max height of the binary search tree.
+  # In a best case scenario, time complexity would be O(logn)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+
+    if key < current_node.key
+      find_helper(current_node.left, key)
+    else
+      find_helper(current_node.right, key)
+    end
+  end
+
+  # Time Complexity: If tree is balanced, it would be O(logn) n being the size of the tree
+  # In a worst case scenario, the time complexity would be O(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(h) h is the height of the tree
+  # Space Complexity: O(h) h is the height of the tree
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    list << { key: current_node.key, value: current_node.value }
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    return list
+  end
+
+  # Time Complexity: O(h) h is the height of the tree
+  # Space Complexity: O(h) h is the height of the tree
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value }
+    return list
+  end
+
+  # Time Complexity: O(h) h is the height of the tree
+  # Space Complexity: O(h) h is the height of the tree
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def height_helper(current_node, max, count)
+    return max if current_node.nil?
+
+    if count > max
+      max = count
+    end
+
+    height_helper(current_node.left, max, count + 1)
+    height_helper(current_node.right, max, count + 1)
+  end
```

Comment:
⚠️ This method isn't working since it's always returning the right subtree's height.